src/LinuxTracing/LeafFunctionCallManagerTest.cpp

```diff
@@ -0,0 +1,338 @@
+
+  static inline unwindstack::MapInfo kNonExecutableMapInfo{
+      nullptr, kNonExecutableMapsStart, kNonExecutableMapsEnd, 0, PROT_READ, kNonExecutableName};
+
+  static inline unwindstack::FrameData kFrame1{
+      .pc = kTargetAddress1,
+      .function_name = "foo",
+      .function_offset = 0,
+      .map_name = kTargetName,
+  };
+
+  static inline unwindstack::FrameData kFrame2{
+      .pc = kTargetAddress2,
+      .function_name = "bar",
+      .function_offset = 0,
+      .map_name = kTargetName,
+  };
+
+  static inline unwindstack::FrameData kFrame3{
+      .pc = kTargetAddress3,
+      .function_name = "baz",
+      .function_offset = 0,
+      .map_name = kTargetName,
+  };
+
+  static inline unwindstack::FrameData kNonExecutableFrame{
+      .pc = kNonExecutableMapsStart,
+      .function_name = "???",
+      .function_offset = 0,
+      .map_name = kNonExecutableName,
+  };
+};
+
+}  // namespace
+
+TEST_F(LeafFunctionCallManagerTest, PatchLeafFunctionCallerReturnsErrorOnSmallStackSamples) {
+  constexpr uint32_t kPid = 10;
+  constexpr uint64_t kStackSize = 13;
+
+  std::vector<uint64_t> callchain;
+  callchain.push_back(kKernelAddress);
+  callchain.push_back(kTargetAddress1);
+  // Increment by one as the return address is the next address.
+  callchain.push_back(kTargetAddress2 + 1);
+  callchain.push_back(kTargetAddress3 + 1);
+
+  CallchainSamplePerfEvent event{callchain.size(), kStackSize};
+  perf_event_sample_id_tid_time_streamid_cpu sample_id{
+      .pid = kPid,
+      .tid = 11,
+      .time = 15,
+      .stream_id = 12,
+      .cpu = 0,
+      .res = 0,
+  };
+  event.ring_buffer_record.sample_id = sample_id;
+  event.ips = callchain;
+
+  event.regs.bp = 2 * SAMPLE_STACK_USER_SIZE_512BYTES;
+  event.regs.sp = 0;
+
+  EXPECT_EQ(Callstack::kFramePointerDwarfStackTooSmallError,
+            leaf_function_call_manager_.PatchLeafFunctionCaller(&event, &maps_, &unwinder_));
+  EXPECT_THAT(event.ips, ElementsAreArray(callchain));
+}
+
+TEST_F(LeafFunctionCallManagerTest, PatchLeafFunctionCallerReturnsErrorOnUnwindingErrors) {
+  constexpr uint32_t kPid = 10;
+  constexpr uint64_t kStackSize = 13;
+
+  std::vector<uint64_t> callchain;
+  callchain.push_back(kKernelAddress);
+  callchain.push_back(kTargetAddress1);
+  // Increment by one as the return address is the next address.
+  callchain.push_back(kTargetAddress2 + 1);
+  callchain.push_back(kTargetAddress3 + 1);
+
+  CallchainSamplePerfEvent event{callchain.size(), kStackSize};
+  perf_event_sample_id_tid_time_streamid_cpu sample_id{
+      .pid = kPid,
+      .tid = 11,
+      .time = 15,
+      .stream_id = 12,
+      .cpu = 0,
+      .res = 0,
+  };
+  event.ring_buffer_record.sample_id = sample_id;
+  event.ips = callchain;
+  event.regs.bp = SAMPLE_STACK_USER_SIZE_512BYTES;
+  event.regs.sp = 10;
+
+  EXPECT_CALL(maps_, Get).WillRepeatedly(Return(nullptr));
+
+  // Unwinding errors could result in empty callstacks:
```

Comment:
I agree with supporting (and testing) this but this comment confuses me a bit as I think that in general we don't expect empty callstacks from libunwindstack (we should always get at least the IP).

Reply:
Rephrased the comment.